bokehjs/src/coffee/models/widgets/data_table.coffee

```diff
@@ -77,10 +74,13 @@ export class DataProvider
       cols = [["index", 1]]
 
     records = @getRecords()
-    records.sort (record1, record2) ->
+    old_index = @index.slice()
+
+    # TODO (bev) this sort is unstable, which is not great
+    @index.sort (i1, i2) ->
       for [field, sign] in cols
-        value1 = record1[field]
-        value2 = record2[field]
+        value1 = records[old_index.indexOf(i1)][field]
+        value2 = records[old_index.indexOf(i2)][field]
```

Comment:
Slightly worried about the scaling performance of this inside the sort comparator. If there's a better way I'd love to know.